design/recovery-internals.md

```diff
@@ -0,0 +1,142 @@
+
+
+## Phase 1: READING_CSTATE
+
+This phase reads the coordinated state (cstate) from coordinators. The cstate includes the DBCoreState structure which describes the transaction systems (such as tLogs and tLogs’ configuration, logRouterTags, txsTags) that exist before the recovery. The coordinated state can have multiple generations of tLogs. 
+
+The transaction system states before the recovery is the starting point for the current recovery to construct the configuration of the next-generation transaction system. Note FDB’s transaction system’s generation increases for each recovery.
+
+
+## Phase 2: LOCKING_CSTATE
+
+This phase locks the coordinated state (cstate) to make sure there is only one master who can change the cstate. Otherwise, we may end up with more than 1 master accepting commits after the recovery. To achieve that, the master needs to get currently alive tLogs’ interfaces and sends commands to tLogs to lock their states, preventing them from accepting any further writes. 
+
+
+FDB lets the current alive tLogs to use the master’s interface to register each tLog’s interface to master. Master simply waits on receiving the TLogRejoinRequest streams: for each tLog’s interface received, the master compares the interface id with the tLog id read from cstate. Once the master collects enough old tLog interfaces, it will use the interfaces to lock those tLogs.
+The logic of collecting tLogs’ interfaces is implemented in trackRejoins() function.
+The logic of locking the tLogs is implemented in epochEnd() function in TagPartitionedLogSystems.actor.cpp.
+
+
+Once we lock the cstate, we bump the recoveryCount and write the cstate to kill the other recovery attempts. This makes the other attempts know they are died and do not recruit more tLogs. If we do not do this, there will be many recovery attempts which recruit tLogs and can make system out of memory. This operation is the reason why every recovery bumps the recoveryCount (epoch) by 2 instead of 1.
+
+
+*How does each tLog know the current master’s interface?*
+Master interface is stored in ServerDBInfo. Once the CC recruits the master, it updates the ServerDBInfo with the master’s interface. CC will send the updated serverDBInfo, which has the master’s interface, to all processes. tLog processes (i,e., tLog workers) monitor the serverDBInfo in an actor. when the serverDBInfo changes, it will register itself to the new master. The logic for a tLog worker to monitor serverDBInfo change is implemented in monitorServerDBInfo() actor.
+
+
+*How does each role, such as tLog and data distributor (DD), register its interface to master and CC?*
+tLog monitors ServerDBInfo change and sends its interface to the new master;
+DD and RateKeeper rejoin themselves to CC because they are no longer a part of the recovery process;
+SS does not rejoin. It waits for the tLogs to be ready and commit their interfaces into database with a special transaction.
+
```

Comment:
Good explanation.
This seems like the first instance of term SS. Can expand it (i.e. clarify we are talking about storage server). Might be worthwhile mentioning the function name where SS waits. Similar to how you have added functions names above.